File: day6/day6.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <set>


int n_unique(std::string str)
{
    std::set<char> str_unique_values(str.begin(), str.end());
    int num_unique = str.length() - (str.length() - str_unique_values.size());
    return num_unique;
}

class DataStreamBuffer
{
private:
    std::string buffer;
    int start_of_marker(int num_distinc_chars);
public:
    DataStreamBuffer(std::string buffer);
    ~DataStreamBuffer();
    int start_of_packet_marker();
    int start_of_message_marker();
};

DataStreamBuffer::DataStreamBuffer(std::string buffer)
{
    this->buffer = buffer;
}

DataStreamBuffer::~DataStreamBuffer(){}

int DataStreamBuffer::start_of_packet_marker()
{
    return this->start_of_marker(4);
}


int DataStreamBuffer::start_of_message_marker()
{
    return this->start_of_marker(14);
}

int DataStreamBuffer::start_of_marker(int num_distinc_chars)
{   
    for (int i = num_distinc_chars; i < this->buffer.length(); i++)
    {
        if (n_unique(this->buffer.substr(i - num_distinc_chars, num_distinc_chars)) == num_distinc_chars)
            return i;
    }
    return 0;    
}

int main()
{   
    std::ifstream filestream("input.txt");
    std::string task_input;
    std::getline(filestream, task_input);
    DataStreamBuffer dsb(task_input);

    std::cout << "PART 1: " << dsb.start_of_packet_marker() << std::endl;
    std::cout << "PART 2: " << dsb.start_of_message_marker() << std::endl;

    filestream.close();
    return 0;
}

File: day6/day6.py
from InputBase import InputBase
from functools import cached_property
from datetime import datetime

class Input(InputBase):
    def read_task_input(self):
        return self.raw_input


def n_unique(iterable):
    num_unique = len(iterable) - (len(iterable) - len(set(iterable)))
    return num_unique


class DataStreamBuffer:
    def __init__(self, buffer: str):
        self.buffer = buffer

    @cached_property
    def start_of_packet_marker(self):
        return self.start_of_marker(4)

    @cached_property
    def start_of_message_marker(self):
        return self.start_of_marker(14)

    def start_of_marker(self, num_distinct_chars):
        for i in range(num_distinct_chars, len(self.buffer)):
            if n_unique(self.buffer[i-num_distinct_chars:i]) == num_distinct_chars:
                return i


def test_pt1():
    tests = [('bvwbjplbgvbhsrlpgdmjqwftvncz', 5),
             ('nppdvjthqldpwncqszvftbrmjlhg', 6),
             ('nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg', 10),
             ('zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw', 11)]
    assert all([DataStreamBuffer(test_buffer).start_of_packet_marker == expected_result
               for test_buffer, expected_result in tests])
    print('PART1 TESTS OK')


def test_pt2():
    tests = [('mjqjpqmgbljsphdztnvjfqwrcgsmlb', 19),
             ('bvwbjplbgvbhsrlpgdmjqwftvncz', 23),
             ('nppdvjthqldpwncqszvftbrmjlhg', 23),
             ('nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg', 29),
             ('zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw', 26)]
    assert all([DataStreamBuffer(test_buffer).start_of_message_marker == expected_result]
               for test_buffer, expected_result in tests)
    print('PART2 TESTS OK')


if __name__ == '__main__':
    start = datetime.now()
    test_pt1()
    test_pt2()

    task_input = Input('input.txt').read_task_input()
    dsb = DataStreamBuffer(task_input)

    print(f'PART 1: {dsb.start_of_packet_marker}')
    print(f'PART 2: {dsb.start_of_message_marker}')

    end = datetime.now()
    print(f'time elapsed: {(end-start).microseconds / 1000}ns')